src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,197 @@
+{
+    internal partial class TemporaryStorageServiceFactory
+    {
+        private sealed class ReferenceCountedDisposable<T> : IDisposable
+            where T : IDisposable
+        {
+            private T _instance;
+            private StrongBox<int> _referenceCount;
+
+            /// <summary>
+            /// Initializes a new reference counting wrapper around an <see cref="IDisposable"/> object.
+            /// </summary>
+            /// <remarks>
+            /// <para>The reference count is initialized to 1.</para>
+            /// </remarks>
+            /// <param name="instance">The object owned by this wrapper.</param>
+            public ReferenceCountedDisposable(T instance)
+            {
+                _instance = instance;
+                _referenceCount = new StrongBox<int>(1);
+            }
+
+            private ReferenceCountedDisposable(T instance, StrongBox<int> referenceCount)
+            {
+                _instance = instance;
+
+                // The reference count has already been incremented for this instance
+                _referenceCount = referenceCount;
+            }
+
+            /// <summary>
+            /// Gets the target object.
+            /// </summary>
+            /// <remarks>
+            /// <para>This call is not valid after <see cref="Dispose"/> is called.</para>
+            /// </remarks>
+            public T Target => _instance;
+
+            /// <summary>
+            /// Increments the reference count for the disposable object, and returns a new disposable reference to it.
+            /// </summary>
+            /// <remarks>
+            /// <para>The returned object is an independent reference to the same underlying object. Disposing of the
+            /// returned value multiple times will only cause the reference count to be decreased once.</para>
+            /// </remarks>
+            /// <returns>A new <see cref="ReferenceCountedDisposable{T}"/> pointing to the same underlying object, if it
+            /// has not yet been disposed; otherwise, <see langword="null"/> if this reference to the underlying object
+            /// has already been disposed.</returns>
+            public ReferenceCountedDisposable<T> TryAddReference()
+            {
+                // This value can be set in Dispose. However, even if we have a race condition with Dispose where it's
+                // set to null in the middle of the call, the reference count will be incremented by this method if and
+                // only if a valid new reference was created. This is because the validity of this instance for creating
+                // new references is determined by _referenceCount.
+                var target = _instance;
+                Interlocked.MemoryBarrier();
+                var referenceCount = Volatile.Read(ref _referenceCount);
+
+                // Cannot use Interlocked.Increment because we need to latch the reference count when it reaches 0.
```

Comment:
No, `CountdownEvent` is not safe for use in concurrent code due to its implementation of `Dispose`. We would have to intentionally fail to call `Dispose` on it (not unlike Microsoft/vs-threading#105). It also adds a bunch of overhead for features we don't need here.